pkg/entrypoint/run.go

```diff
@@ -0,0 +1,295 @@
+	InternalErrorCode = 127
+	// AbortedErrorCode is what we write to the marker file to
+	// indicate that we were terminated via a signal.
+	AbortedErrorCode = 130
+
+	// DefaultTimeout is the default timeout for the test
+	// process before SIGINT is sent
+	DefaultTimeout = 120 * time.Minute
+
+	// DefaultGracePeriod is the default timeout for the test
+	// process after SIGINT is sent before SIGKILL is sent
+	DefaultGracePeriod = 15 * time.Second
+
+	// DefaultShouldWaitForPrevStep is ...
+	DefaultShouldWaitForPrevStep = false
+
+	// DefaultShouldRunPostRun is ...
+	DefaultShouldRunPostRun = false
+
+	// DefaultPreRunFile is ...
+	DefaultPreRunFile = "0"
+
+	// DefaultPostRunFile is ...
+	DefaultPostRunFile = "1"
+)
+
+var (
+	// errTimedOut is used as the command's error when the command
+	// is terminated after the timeout is reached
+	errTimedOut = errors.New("process timed out")
+	// errAborted is used as the command's error when the command
+	// is shut down by an external signal
+	errAborted = errors.New("process aborted")
+)
+
+// Run executes the test process then writes the exit code to the marker file.
+// This function returns the status code that should be passed to os.Exit().
+func (o Options) Run() int {
+	code, err := o.ExecuteProcess()
+	if err != nil {
+		logrus.WithError(err).Error("Error executing test process")
+	}
+	if err := o.mark(code); err != nil {
+		logrus.WithError(err).Error("Error writing exit code to marker file")
+		return InternalErrorCode
+	}
+	return code
+}
+
+// ExecuteProcess creates the artifact directory then executes the process as
+// configured, writing the output to the process log.
+func (o Options) ExecuteProcess() (int, error) {
+	if o.ArtifactDir != "" {
+		if err := os.MkdirAll(o.ArtifactDir, os.ModePerm); err != nil {
+			return InternalErrorCode, fmt.Errorf("could not create artifact directory(%s): %v", o.ArtifactDir, err)
+		}
+	}
+	processLogFile, err := os.Create(o.ProcessLog)
+	if err != nil {
+		return InternalErrorCode, fmt.Errorf("could not create process logfile(%s): %v", o.ProcessLog, err)
+	}
+	defer processLogFile.Close()
+
+	output := io.MultiWriter(os.Stdout, processLogFile)
+	logrus.SetOutput(output)
+	defer logrus.SetOutput(os.Stdout)
+
+	// ---
+	timeout := optionOrDefault(o.Timeout, DefaultTimeout)
+	gracePeriod := optionOrDefault(o.GracePeriod, DefaultGracePeriod)
+	var commandErr error
+	cancelled, aborted := false, false
+
+	done := make(chan error)
+	go func() {
+		done <- o.waitForPrevStep()
+	}()
+	select {
+	case err := <-done:
+		commandErr = err
+		if err != nil {
+			cancelled = true
+		}
+	case <-time.After(timeout):
+		logrus.Errorf("Previous step did not finish within %s timeout", timeout)
+		cancelled = true
+	}
+
+	var returnCode int
+	if cancelled {
+		if aborted {
+			commandErr = errAborted
+			returnCode = AbortedErrorCode
+		} else {
+			commandErr = errTimedOut
+			returnCode = InternalErrorCode
+		}
+		return returnCode, commandErr
+	}
+
+	// ---
```

Comment:
what do these `// ---` mean? (maybe they are good points at which to consider breaking this up into separate functions?)